feature(nemesis): introduce nemesis node allocator

dimakr opened the pull request:
Previously we were adding locked sections in different parts of nemesis.py, to prevent parallel nemeses from selecting the same target node. However, managing these locks across multiple locations became increasingly difficult, but race condition issue appears periodically in new places.

This change introduces a NemesisNodeAllocator entity, which centralizes the logic of selecting and tracking nodes used in nemeses execution. It helpts to ensure synchronization between parallel nemeses.

### Testing
<!-- Add links to Argus/Jenkins of test test done with this PR -->
<!-- This would help the reviewer to cross check what was tested, and and review the results as needed -->
- [x] [longevity-schema-topology-changes-12h-test tier1 longevity, that has parallel nemeses](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/longevity-schema-topology-changes-12h-test/8/) (the step of logs collection failed in this run as builder instance went offline at the end, but the test itself was executed and node allocator logs are also present in Jenkins console output)
- [x] check for parallel nemesis synchronization of target node selection: [longevity-100gb-4h with 2 parallel nemeses](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-100gb-4h-test/20/)
Some SCT log entries from the new node_allocator module:
```
❯ egrep -i 'node_allocator' sct-f3405cf3.log
< t:2025-06-21 20:16:23,417 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > StartStopMajorCompaction-c86c920b: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-5. Marked it as running nemesis.
< t:2025-06-21 20:16:23,445 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > StartStopMajorCompaction-f3c05dc5: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-4. Marked it as running nemesis.
< t:2025-06-21 20:16:35,462 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > StartStopMajorCompaction-f3c05dc5: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-4']
< t:2025-06-21 20:16:36,573 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > StartStopMajorCompaction-c86c920b: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-5']
< t:2025-06-21 20:21:35,463 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SlaIncreaseSharesDuringLoad-83d38987: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-5. Marked it as running nemesis.
< t:2025-06-21 20:21:36,574 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SlaIncreaseSharesDuringLoad-211f76a4: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-4. Marked it as running nemesis.
< t:2025-06-21 20:21:37,265 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SlaIncreaseSharesDuringLoad-83d38987: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-5']
< t:2025-06-21 20:21:37,267 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > CreateIndex-91b81680: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-2. Marked it as running nemesis.
< t:2025-06-21 20:21:38,381 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SlaIncreaseSharesDuringLoad-211f76a4: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-4']
< t:2025-06-21 20:21:38,386 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > CreateIndex-02a0f65a: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-1. Marked it as running nemesis.
< t:2025-06-21 20:21:38,838 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > CreateIndex-91b81680: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-2']
< t:2025-06-21 20:21:38,839 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > StartStopCleanupCompaction-4e7e9c22: selected target node longevity-100gb-4h-sync-par-db-node-f3405cf3-4. Marked it as running nemesis.
< t:2025-06-21 20:21:39,949 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > CreateIndex-02a0f65a: unset running nemesis from nodes: ['longevity-100gb-4h-sync-par-db-node-f3405cf3-1']
...
```
- [x] check for `switch target node` capability: [longevity-5gb-1h-nemesis + SerialRestartOfElectedTopologyCoordinatorNemesis test configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/longevity-100gb-4h-test/21/)
Some SCT log entries from the new node_allocator module, which indicate that switching was performed:
```
❯ egrep -i 'node_allocator' sct-2da3077f.log
< t:2025-06-22 11:33:52,596 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-c2d13129: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1. Marked it as running nemesis.
< t:2025-06-22 11:33:54,357 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-8de7f587: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3. Marked it as running nemesis.
< t:2025-06-22 11:33:55,718 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-8de7f587: unset running nemesis for node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3.
< t:2025-06-22 11:34:11,918 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-5cbd5eb8: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-2. Marked it as running nemesis.
< t:2025-06-22 11:34:13,286 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-5cbd5eb8: unset running nemesis for node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-2.
< t:2025-06-22 11:34:39,711 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-c2d13129: unset running nemesis from nodes: ['longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1']
< t:2025-06-22 11:37:55,230 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-7f84e26a: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1. Marked it as running nemesis.
< t:2025-06-22 11:37:56,800 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-f45484dd: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3. Marked it as running nemesis.
< t:2025-06-22 11:37:58,653 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-f45484dd: unset running nemesis for node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3.
< t:2025-06-22 11:37:58,653 f:node_allocator.py l:197  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-7f84e26a: released old target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1 during switch.
< t:2025-06-22 11:37:58,653 f:node_allocator.py l:205  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-7f84e26a: switched target node to longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-2.
< t:2025-06-22 11:38:19,850 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-1e1aa89d: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1. Marked it as running nemesis.
< t:2025-06-22 11:38:21,717 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-1e1aa89d: unset running nemesis for node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1.
< t:2025-06-22 11:38:54,050 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-7f84e26a: unset running nemesis from nodes: ['longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-2']
< t:2025-06-22 11:42:09,661 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-31705a0c: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1. Marked it as running nemesis.
< t:2025-06-22 11:42:11,222 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-cf063712: selected target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3. Marked it as running nemesis.
< t:2025-06-22 11:42:13,583 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SearchCoordinator-cf063712: unset running nemesis for node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3.
< t:2025-06-22 11:42:13,583 f:node_allocator.py l:197  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-31705a0c: released old target node longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-1 during switch.
< t:2025-06-22 11:42:13,583 f:node_allocator.py l:205  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > SerialRestartElectedTopologyCoordinator-31705a0c: switched target node to longevity-5gb-1h-SerialRestartOfEle-db-node-2da3077f-3.
...
```
- [x] check for new nodes being immediately marked as running a nemesis when added during the nemesis execution: [longevity-100gb-4h + GrowShrinkClusterNemesis'](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-100gb-4h-test-copy1/10/)
```
❯ egrep -i 'node_allocator' sct-7c57fa12.log
< t:2025-06-21 18:12:24,609 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-5. Marked it as running nemesis.
< t:2025-06-21 18:13:26,828 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7.
< t:2025-06-21 18:13:26,828 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-8.
< t:2025-06-21 18:13:26,828 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-9.
< t:2025-06-21 18:22:07,242 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: unset running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-5.
< t:2025-06-21 18:22:07,242 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-4. Marked it as running nemesis.
< t:2025-06-21 18:22:07,257 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-2. Marked it as running nemesis.
< t:2025-06-21 18:22:07,271 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-3. Marked it as running nemesis.
< t:2025-06-21 18:29:16,283 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-9bc1447c: unset running nemesis from nodes: ['longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-8', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-9', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-4', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-2', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-3']
< t:2025-06-21 18:34:50,979 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7. Marked it as running nemesis.
< t:2025-06-21 18:35:51,637 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-10.
< t:2025-06-21 18:35:51,637 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-11.
< t:2025-06-21 18:35:51,637 f:node_allocator.py l:145  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: set running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-12.
< t:2025-06-21 18:44:32,654 f:node_allocator.py l:155  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: unset running nemesis for node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7.
< t:2025-06-21 18:44:32,654 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7. Marked it as running nemesis.
< t:2025-06-21 18:44:32,670 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-5. Marked it as running nemesis.
< t:2025-06-21 18:44:32,684 f:node_allocator.py l:126  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: selected target node longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-9. Marked it as running nemesis.
< t:2025-06-21 18:52:25,488 f:node_allocator.py l:174  c:sdcm.utils.nemesis_utils.node_allocator p:INFO  > GrowShrinkCluster-24ba2f63: unset running nemesis from nodes: ['longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-10', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-11', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-12', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-7', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-5', 'longevity-5gb-1h-GrowShrinkClusterN-db-node-7c57fa12-9']
...
```

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

pehala commented on sdcm/cluster_azure.py:
```diff
@@ -229,6 +229,10 @@ def add_nodes(self, count, ec2_user_data='', dc_idx=0, rack=0, enable_auto_boots
             self.log.info("Added node: %s", node.name)
             node.enable_auto_bootstrap = enable_auto_bootstrap
 
+            if self.nemesis_allocator and 'db' in node.node_type:
```
the identical code is added to all clusters, cannot we add it somewhere only once?

dimakr replied:
Removed this from different cluster backends, after added property to get all nodes from all clusters in tester object

pehala replied:
Thanks, much better looking (and more consise as well)

pehala commented on sdcm/utils/nemesis_utils/common.py:
I would make this `__init__.py` due to low amount of actual code

dimakr replied:
Moved to __init__

pehala commented on sdcm/utils/nemesis_utils/common.py:
I would use `utils/nemesis` for package name and moved all nemesis related utils there (like nemesis.py and nemesis_jobs_configs)

dimakr replied:
I suggest to do this refactoring in a separate change. It is out of scope of this task; and we can think of what else can be moved out of main nemesis.py module, or other places, to utils

pehala replied:
The issue is that you introduce package with a bad name in my opinion, just because the correct name is already taken here (`utils/nemesis`). If we do the refactor directly after, I am ok with that, but I worry we will never get to it, which I suspect is the reason why we ended with the current messy structure

pehala replied:
I have PR (https://github.com/scylladb/scylla-cluster-tests/pull/10385) which as one of the changes renames the `utils/nemesis.py` to `utils/nemesis_generation.py`, which would open possibility to introduce the `utils/nemesis` package here and other things to move there in followups

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,174 @@
+                        nemesis_name, selected_node.name)
+            return selected_node
+
+    def notify_node_added(self, node: BaseNode):
```
The fact the we have to also manage nodes here (compared to all different places we do it already), tell me we should revisit design and make sure we need to add nodes only once. This way it could be easily forgotten or out of sync

dimakr replied:
`tester.db_cluster.nodes` is aware of nodes of only one cluster, but for for multitenant case it won't be enough.
I.e. we would need for the allocator to know abot all nodes from all cluster of tester.db_clusters_multitenant, if I read this correctly.

dimakr replied:
Probably we can introduce a property in tester object, which will return all nodes from all db_cluster?
This way we don't need to track add/remove of nodes separately

dimakr replied:
Changed this - removed tracking/notifications in node allocator of added/removed nodes. Nodes list directly from tester is used instead.

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,174 @@
+                      dc_idx: int | None = None,
+                      rack: int | None = None,
+                      filter_seed: bool = False) -> list[BaseNode]:
+        if is_seed is DefaultValue:
```
how is None different to DefaultValue here?

dimakr replied:
This is existing logic of `Nemesis._get_target_nodes` copied from nemesis.py to this new module as is. 
I should've copied the doc. string from there as well. Will do this.

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,174 @@
+
+class NemesisNodeAllocator:
+    def __init__(self):
+        self.lock = threading.Lock()
```
If I understand this correctly, this will work only if there is only once instance of this class? If that is correct, we should make this class Singleton

dimakr replied:
This class is instantiated only once, in tester object. So there will be one instance of it during the test run.

pehala replied:
Ok, conceptually, I still think that making this class a singleton is a good idea

pehala commented on sdcm/nemesis.py:
```diff
@@ -377,56 +372,19 @@ def publish_event(self, disrupt, status=True, data=None):
+        else:
+            raise UnsupportedNemesis(
+                f"{self.current_disruption}: failed to switch target node {self.target_node.name} to {node.name} "
+                "(it is already busy).")
```
Busy due to what?

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,188 @@
+                    LOGGER.warning("%s: tried to release node %s, but it was busy with a different nemesis %s.",
+                                   nemesis_name, node.name, self.active_nemeses_on_nodes[node])
+            else:
+                LOGGER.warning("%s: tried to release node %s, but it was not marked as busy by the allocator.",
```
Just a note: We should monitor these closely, as they most likely indicate implementation error

fruch replied:
Hence we should raise as an error event, warning wouldn't be noticed.

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,188 @@
+            return True
+
+    @contextmanager
+    def mark_nodes_as_busy_context(self, nodes: list[BaseNode] | BaseNode, nemesis_name: str):
```
I do not like the naming, but I do not have any better suggestions right now :/

fruch replied:
Maybe something like


with nodes_marked_as_busy(...)

with nodes_marked_as_in_use(...)

Even nodes can be dropped, cause it's the first argument

dimakr replied:
Changed the name

pehala commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,188 @@
+    def switch_target_node(self, old_node: BaseNode, new_node: BaseNode, nemesis_name: str) -> bool:
+        """Switches the nemesis target from an old node to a new one."""
+        if old_node == new_node:
+            LOGGER.debug(
```
Not sure if we need that message, I know it is just a debug, but given the already big spam we do with all of our logs, I do not think we need to add more

dimakr replied:
Removed the message

pehala commented on sdcm/nemesis.py:
```diff
@@ -5255,34 +5201,35 @@ def disrupt_bootstrap_streaming_error(self):
         new_node: BaseNode = skip_on_capacity_issues(self.cluster.add_nodes)(
```
I would add mark_node calls to add_nodes directly, they should be marked right after they spawn. If we for some reason do not want that we can flag to the parameters.

I think this way of doing thing would clear a lot of unneccessary code and responsiblity from nemesis functions and would make it more logical, if the nemesis spawns a node it should automatically be added to that nemesis active nodes.

I suspect the problem with this wopuld be that we are releasing node, so we should do release_all function which could clear all reserved nodes automatically after nemesis ends

fruch replied:
add_nodes doesn't have references to nemesis or the test todo so

and also that would spread out to multiple implementations of add_nodes, which is what was the first approach in this PR.

Yes we should mark a node in use as it introduces into db_cluster.nodes, but it would still be the responsibility of the nemesis code to release it, every code has it's one logic until when it needs the newly introduced node

dimakr replied:
Changed the approach as described above - when nodes are automatically marked as running nemesis and monitored by node allocator, if they were created during nemesis

dimakr commented:
@pehala @fruch
addressed comments from the latest round of review, the changes mainly:
- added a decorator to be used around `cluster.add_nodes` methods for different backends, so that allocator marks the new node as running nemesis, when the nodes are created by a nemesis
- added release of all nodes that tun current_disruption, at the end of a disruption in  Nemesis.run. This is to ensure after anew-nodes are created by nemesis (and automatically reserved) are not left reserved after nemesis is finished
- calls to `cluster.add_nodes` in nemesis.py are properly parametrized, so that new nodes are automatically marked with running nemesis
- raise exceptions explicitly, not just skipping nemesis with UnsupportedNemesis
- some other changes - improved messages in log entries, renamed public APIs in node_allocator object (`busy` naming became clunky, from my pov), etc.

Now this change is being tested

fruch commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,247 @@
+        be able to use the new node(s), we need to release them after disruption is finished.
+        """
+        with self.lock:
+            nodes_to_release = [node for node, owner in self.active_nemeses_on_nodes.items() if owner == nemesis_name]
```
I'm afraid name of nemesis is enough, what happens if two nemesis threads are running the exact same nemesis ?

fruch commented:
@dimakr one more conflict

dimakr commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,247 @@
+    def __init__(self, tester_obj):
+        self.lock = threading.Lock()
+        self._tester = tester_obj
+        self.active_nemeses_on_nodes: dict['BaseNode', str] = {}
```
Not really a typo (i used plural form of nemesis to indicate that this attribute will hold many disruption items)
But I agree that we should use the same name everywhere in the code.
Fixed it.

dimakr commented on sdcm/utils/nemesis_utils/node_allocator.py:
```diff
@@ -0,0 +1,247 @@
+    @contextmanager
+    def nodes_running_nemesis(self, nodes: Union[list['BaseNode'], 'BaseNode'], nemesis_name: str):
+        """Temporarily marks nodes as running the specified nemesis."""
+        from sdcm.cluster import BaseNode  # importing here to avoid circular import
+        if isinstance(nodes, BaseNode):
+            nodes = [nodes]
```
Changed to checking if nodes are an iterable.

dimakr commented:
@vponomaryov @fruch I addressed your latest comments
when you have a minute, can you check if this change is OK to move forward?

dimakr commented:
@fruch I don't remember if we discussed backporting this change anywhere, but in the latest  2024.1 patch release we again ran into racy problem of 2 parallel nemeses selecting the same node.
Maybe we can plan backporting this (and run_nemesis fix) after we have this executed on the next weekly run.

fruch commented:
> @fruch I don't remember if we discussed backporting this change anywhere, but in the latest  2024.1 patch release we again ran into racy problem of 2 parallel nemeses selecting the same node.
> Maybe we can plan backporting this (and run_nemesis fix) after we have this executed on the next weekly run.

Not sure 2024.1 dim such a big backport... (Only few months left for it support)
